config.md
# Configuration

The configuration is stored in `config/adaptiveview.json`. A default config is generated when the server is first started with the mod. Otherwise you can find the default configuration [here](default_config.json).

## Basic Configuration

### Global rules

These values apply if there are no rules active that override them.

- `update_rate`: The interval in ticks after which performance is checked and view distance is adjusted.
- `max_view_distance`: The maximum view distance the dynamic adjustment can go up to.
- `min_view_distance`: The minimum view distance the dynamic adjustment can go down to.
- `max_sim_distance`: The maximum simulation distance the dynamic adjustment can go up to.
- `min_sim_distance`: The minimum simulation distance the dynamic adjustment can go down to.
- `max_chunk_ticking_distance`: The maximum chunk-ticking distance the dynamic adjustment can go up to.
- `min_chunk_ticking_distance`: The minimum chunk-ticking distance the dynamic adjustment can go down to.

### Other Configurations
- `allow_on_client`: If this is enabled the mod will try to work in a client only environment. This can lead to unexpected behaviour.
- `broadcast_changes_default`: Sets which player group (`NONE`, `OPS`, `ALL`) gets notified when the view distance changes by default.
- `broadcast_changes`: List of (case-insensitive) player names, that get notified when the view distance changes. If the name is prefixed with `!`, the player will not get notified..
- `broadcast_lock_default`: Sets which player group (`NONE`, `OPS`, `ALL`) gets notified when the view distance is locked or unlocked by default.
- `broadcast_lock`: List of (case-insensitive) player names, that get notified when the view distance is locked or unlocked. If the name is prefixed with `!`, the player will not get notified. 

## Rules

Rules specify by which conditions view distance is adjusted. They are made up of a condition and an action that is applied when the condition is met.

Rules also support custom names using the `name` property.

### Condition
- `type`: One of `MSPT`, `MEMORY` or `PLAYERS`. The kind of condition this is.

All the following need to be met for the rule to be active:

- `min`: The minimum MSPT, Memory usage (in percent) or online Players required for the rule to be active.
- `max`: The maximum MSPT, Memory usage (in percent) or online players allowed for the rule to be active.
- `value`: A list of (case-insensitive) player names (only active if `type` is `PLAYERS`) separated by `,`. The string can be prefixed with the following values to specify when the rule is active:
  - No prefix: Any of the specified players are online
  - `&`: All of the specified players are online
  - `!`: Any not specified player is online
  - `\`: None of the specified players are online

### Action
- `target`: One of `VIEW`, `SIMULATION` or `CHUNK_TICKING`. Whether view-, simulation- or chunk-ticking distance is affected.

> [!WARNING]
> Due to the way Minecraft handles chunk tracking, blocks outside of an 17x17 chunk area around the player will not receive chunk ticks, no matter what the `CHUNK_TICKING` rules set. 
> This probably cannot be changed without rewriting large parts of chunk loading code.

If the condition is met, the following actions are performed:

- `update_rate`: Overrides the global `update_rate` value.
- `max_distance`: Overrides the global `max_view_distance`, `max_sim_distance` or `max_chunk_ticking_distance` value depending on `target`.
- `min_distance`: Overrides the global `minv_iew_distance`, `min_sim_distance` or `min_chunk_ticking_distance` value depending on `target`.
- `step`: The amount of chunks the target is changed by every update. Positive values increase the  distance and negative values decrease it. 
- `step_after`: The amount of update after which the step is performed. E.g. if `update_rate` is `100` and `step_after` is `2`, the `step` will be performed every `200` ticks.

### Resolving conflicts

If the same value is set by multiple active rules, the following way of resolving conflicts is used:

- `update_rate`: The lowest value is used.
- `min_distance`: The highest value is used.
- `max_distance`: The lowest value is used. If `min_distance` is greater than `max_distance`, the value of `min_distance` is used.
- `step`: If there are negative step values, the lowest one is used, else the highest one is used.
- `step_after`: Individual per rule. Effects the rule's `step`.

In general, the value that is more limiting is used.

## Examples

### The default config

A basic config applicable to most servers struggling for MSPT performance.

1. The View Distance is updated every 600 ticks (30 seconds).
2. The View Distance is allowed to be between 4 and 20 chunks.
3. If the server is between 40 and 50 MSPT, nothing happens.
4. If the server is above 50 MSPT, the view distance is adjusted down by 1 chunk every update. 
5. If the server is above 60 MSPT, the view distance is adjusted down by 2 chunks every update.
6. If the server is below 40 MSPT, the view distance is adjusted up by 1 chunk every update.
7. If the server is below 30 MSPT, the view distance is adjusted up by 2 chunks every update.



```json
{
  "update_rate": 600, // 1.
  "max_view_distance": 20, // 2.
  "min_view_distance": 4, // 2.
  [...]
  "rules": [
    {
      "type": "MSPT",
      "min": 60,
      "target": "VIEW",
      "step": -2  // 5.
    },
    {
      "type": "MSPT",
      "min": 50,
      "target": "VIEW",
      "step": -1  // 4.
    },
    {
      "type": "MSPT",
      "max": 40,
      "target": "VIEW",
      "step": 1  // 6.
    },
    {
      "type": "MSPT",
      "max": 30,
      "target": "VIEW",
      "step": 2  // 7.
    }
  ]
}
```

### Memory focused config

A config that could be used, if a server is low on memory.

> [!NOTE] 
> Updating view distance based on memory too quickly can be problematic, as it can take a while for memory usage to change after decreasing view distance.

1. The View Distance is updated every 1200 ticks (1 mínute),
2. If there is at most one player online, the View Distance is at most 28 chunks.
3. If there are 2 players online, the View Distance is at most 15 chunks. The update rate is also increased to 300 ticks to allow for closer monitoring of memory usage.
4. If there are 3 or more players online, the View Distance is at most 12 chunks. The update rate of 300 ticks is still active from 3.
5. If the memory usage is at least 90%, the view distance is adjusted down by 1 every 2 minutes (2400 ticks).

```json
{
  "update_rate": 1200, // 1.
  "max_view_distance": 18, // 2.
  "min_view_distance": 4,
  [...]
  "rules": [
    {
      "type": "PLAYERS",
      "min": 2,
      "target": "VIEW",
      "max_distance": 15, // 3.
      "update_rate": 300  // 3.
    },
    {
      "type": "PLAYERS",
      "min": 3,
      "target": "VIEW",
      "max_distance": 12, // 4.
    },
    {
      "type": "MEMORY",
      "min": 90,
      "target": "VIEW",
      "step": -1,  // 5.
      "step_after": 8,  // 5.
      "update_rate": 300
    }
  ]
}
```

### Config that locks the view distance if specific players are online

This config may e.g. be useful if you have specific Alt-Accounts that are always used for running farms. This also works if the player is a carpet bot.

1. Everything works just like the default config.
2. If the player `FarmPlayer1` or `FarmPlayer2` is online, the view distance is locked to 6 chunks and the update rate is decreased to 5 Minutes, as nothing is changing.

```json
{
  "update_rate": 600,
  "max_view_distance": 20,
  "min_view_distance": 4,
  [...]
  "rules": [  
    {
      "type": "PLAYERS",
      "value": "farmplayer1,farmplayer2",
      "target": "VIEW",
      "max_distance": 6,  // 2.
      "min_distance": 6,  // 2.
      "update_rate": 6000  // 2.
    },
    {
      "type": "MSPT",
      "min": 60,
      "target": "VIEW",
      "step": -2
    },
    {
      "type": "MSPT",
      "min": 50,
      "target": "VIEW",
      "step": -1
    },
    {
      "type": "MSPT",
      "max": 40,
      "target": "VIEW",
      "step": 1
    },
    {
      "type": "MSPT",
      "max": 30,
      "target": "VIEW",
      "step": 2
    }
  ]
}
```

## Configuration Specification

Option names followed by a `?` denote optional options.

```json
{
  "name": string,
  "update_rate": int[1..],
  "max_view_distance": int[2..32],
  "min_view_distance": int[2..32],
  "max_sim_distance": int[2..32],
  "min_sim_distance": int[2..32],
  "max_chunk_ticking_distance": int[2..32],
  "min_chunk_ticking_distance": int[2..32],
  "allow_on_clinet": boolean,
  "broadcast_changes_default": string[NONE/OPS/ALL],
  "broadcast_changes": [
    string
  ],
  "broadcast_lock_default": string[NONE/OPS/ALL],
  "broadcast_lock": [
    string
  ],
  "rules": [
    {
      "type": string[MSPT/MEMORY/PLAYERS],
      "value?": string,
      "max?": int[0..],
      "max?": int[0..],
      "target": string[VIEW/SIMULATION/CHUNK_TICKING],
      "update_rate?": int[1..],
      "max_distance?": int[2..32],
      "min_distance?": int[2..32],
      "step?": int,
      "step_after?": int[1..]
    }
  ]
}
```


src/main/java/net/treset/adaptiveview/config/Config.java
package net.treset.adaptiveview.config;

import com.google.gson.*;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;
import net.treset.adaptiveview.AdaptiveViewMod;
import net.treset.adaptiveview.tools.BroadcastLevel;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;

public class Config {
    private static final File configFile = new File("./config/adaptiveview.json");
    private static final Gson gson = new GsonBuilder()
            .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
            .setPrettyPrinting()
            .create();

    private int updateRate;
    private int maxViewDistance;
    private int minViewDistance;
    private int maxSimDistance;
    private int minSimDistance;
    private int maxChunkTickingDistance;
    private int minChunkTickingDistance;
    private boolean allowOnClient;
    @SerializedName("broadcast_to_ops")
    @Expose(serialize = false)
    private boolean deprecatedBroadcastToOps;
    private BroadcastLevel broadcastChangesDefault;
    @SerializedName(value = "broadcast_changes", alternate = "broadcast_to")
    private ArrayList<String> broadcastChanges;
    private BroadcastLevel broadcastLockDefault;
    private ArrayList<String> broadcastLock;
    private ArrayList<Rule> rules;

    private transient boolean viewLocked = false;
    private transient boolean simLocked = false;
    private transient boolean chunkTickingLocked = false;

    public Config(int updateRate, int maxViewDistance, int minViewDistance, int maxSimDistance, int minSimDistance, int maxChunkTickingDistance, int minChunkTicking, boolean allowOnClient, BroadcastLevel broadcastChangesDefault, ArrayList<String> broadcastChanges, BroadcastLevel broadcastLockDefault, ArrayList<String> broadcastLock, ArrayList<Rule> rules) {
        this.updateRate = updateRate;
        this.maxViewDistance = maxViewDistance;
        this.minViewDistance = minViewDistance;
        this.maxSimDistance = maxSimDistance;
        this.minSimDistance = minSimDistance;
        this.maxChunkTickingDistance = maxChunkTickingDistance;
        this.minChunkTickingDistance = minChunkTicking;
        this.allowOnClient = allowOnClient;
        this.rules = rules;
        this.broadcastChangesDefault = broadcastChangesDefault;
        this.broadcastChanges = broadcastChanges;
        this.broadcastLockDefault = broadcastLockDefault;
        this.broadcastLock = broadcastLock;
    }

    public static Config generic() {
        return new Config(
            600,
            20,
            4,
            20,
            4,
            8,
            2,
            false,
            BroadcastLevel.NONE,
            new ArrayList<>(),
            BroadcastLevel.OPS,
            new ArrayList<>(),
            new ArrayList<>(List.of(
                    new Rule(
                            RuleType.MSPT,
                            null,
                            null,
                            60,
                            RuleTarget.VIEW,
                            null,
                            -2,
                            null,
                            null,
                            null,
                            null
                    ),
                    new Rule(
                            RuleType.MSPT,
                            null,
                            null,
                            50,
                            RuleTarget.VIEW,
                            null,
                            -1,
                            null,
                            null,
                            null,
                            null
                    ),
                    new Rule(
                            RuleType.MSPT,
                            null,
                            40,
                            null,
                            RuleTarget.VIEW,
                            null,
                            1,
                            null,
                            null,
                            null,
                            null

                    ),
                    new Rule(
                            RuleType.MSPT,
                            null,
                            30,
                            null,
                            RuleTarget.VIEW,
                            null,
                            2,
                            null,
                            null,
                            null,
                            null
                    )
            ))
        );
    }

    public static Config loadOrDefault() {
        try  {
            return load();
        } catch (IOException e) {
            return generic();
        }
    }

    public static Config load() throws IOException {
        if(!configFile.exists()) {
            return migrateOldConfig();
        }

        String json;
        try {
            json = Files.readString(configFile.toPath());
        } catch (IOException e) {
            AdaptiveViewMod.LOGGER.warn("Failed to read config, using default", e);
            throw e;
        }

        try {
            Config config = gson.fromJson(json, Config.class);
            if(config.rules == null) {
                AdaptiveViewMod.LOGGER.warn("Config not valid, trying to migrate old one");
                return migrateOldConfig();
            }
            for (Rule rule : config.rules) {
                if(!rule.isEffective()) {
                    AdaptiveViewMod.LOGGER.warn("Rule is not effective: {}", rule);
                }
            }
            amendMigration(config);
            AdaptiveViewMod.LOGGER.info("Loaded config");
            return config;
        } catch (JsonSyntaxException e) {
            AdaptiveViewMod.LOGGER.warn("Failed to parse config, using default", e);
            throw new IOException("Failed to load config", e);
        }
    }

    private static void amendMigration(Config config) {
        if(config.getBroadcastChangesDefault() == null) {
            config.setBroadcastChangesDefault(config.deprecatedIsBroadcastToOps() ? BroadcastLevel.OPS : BroadcastLevel.NONE);
        }
        if(config.getBroadcastChanges() == null) {
            config.setBroadcastChanges(new ArrayList<>());
        }
        if(config.getBroadcastLockDefault() == null) {
            config.setBroadcastLockDefault(BroadcastLevel.OPS);
        }
        if(config.getBroadcastLock() == null) {
            config.setBroadcastLock(new ArrayList<>());
        }
    }

    private static Config migrateOldConfig() {
        OldConfig oldConfig = OldConfig.load();
        if(oldConfig == null) {
            AdaptiveViewMod.LOGGER.info("Creating new config...");
            Config config = Config.generic();
            config.save();
            AdaptiveViewMod.LOGGER.info("Crated new config.");
            return config;
        }

        AdaptiveViewMod.LOGGER.info("Migrating old config...");

        Config config = new Config(
            oldConfig.getUpdateInterval(),
            oldConfig.getMaxViewDistance(),
            oldConfig.getMinViewDistance(),
            oldConfig.getMaxViewDistance(),
            oldConfig.getMinViewDistance(),
            8,
            8,
            oldConfig.isOverrideClient(),
            BroadcastLevel.NONE,
            new ArrayList<>(),
            BroadcastLevel.OPS,
            new ArrayList<>(),
            new ArrayList<>(List.of(
                    new Rule(
                            RuleType.MSPT,
                            null,
                            null,
                            oldConfig.getMaxMsptAggressive(),
                            RuleTarget.VIEW,
                            null,
                            -2,
                            null,
                            null,
                            null,
                            null
                    ),
                    new Rule(
                            RuleType.MSPT,
                            null,
                            null,
                            oldConfig.getMaxMspt(),
                            RuleTarget.VIEW,
                            null,
                            -1,
                            null,
                            null,
                            null,
                            null
                    ),
                    new Rule(
                            RuleType.MSPT,
                            null,
                            oldConfig.getMinMspt(),
                            null,
                            RuleTarget.VIEW,
                            null,
                            1,
                            null,
                            null,
                            null,
                            null

                    ),
                    new Rule(
                            RuleType.MSPT,
                            null,
                            oldConfig.getMinMsptAggressive(),
                            null,
                            RuleTarget.VIEW,
                            null,
                            2,
                            null,
                            null,
                            null,
                            null
                    )
            ))
        );

        config.save();

        AdaptiveViewMod.LOGGER.info("Migrated new config.");
        return config;
    }

    public void copy(Config config) {
        this.updateRate = config.updateRate;
        this.maxViewDistance = config.maxViewDistance;
        this.minViewDistance = config.minViewDistance;
        this.maxSimDistance = config.maxSimDistance;
        this.minSimDistance = config.minSimDistance;
        this.maxChunkTickingDistance = config.maxChunkTickingDistance;
        this.minChunkTickingDistance = config.minChunkTickingDistance;
        this.allowOnClient = config.allowOnClient;
        this.broadcastChangesDefault = config.broadcastChangesDefault;
        this.broadcastChanges = config.broadcastChanges;
        this.broadcastLockDefault = config.broadcastLockDefault;
        this.broadcastLock = config.broadcastLock;
        this.rules = config.rules;
    }

    public void save() {
        if(!configFile.exists()) {
            try {
                Files.createDirectories(configFile.getParentFile().toPath());
                Files.createFile(configFile.toPath());
            } catch (IOException e) {
                AdaptiveViewMod.LOGGER.error("Failed to create config file", e);
            }
        }

        String json = gson.toJson(this);
        try {
            Files.writeString(configFile.toPath(), json);
            AdaptiveViewMod.LOGGER.info("Saved config");
        } catch (IOException e) {
            AdaptiveViewMod.LOGGER.error("Failed to write config file", e);
        }
    }

    public int getUpdateRate() {
        return updateRate;
    }

    public void setUpdateRate(int updateRate) {
        this.updateRate = updateRate;
    }

    public int getMaxViewDistance() {
        return maxViewDistance;
    }

    public void setMaxViewDistance(int maxViewDistance) {
        this.maxViewDistance = maxViewDistance;
    }

    public int getMinViewDistance() {
        return minViewDistance;
    }

    public void setMinViewDistance(int minViewDistance) {
        this.minViewDistance = minViewDistance;
    }

    public int getMaxSimDistance() {
        if(maxSimDistance == 0) {
            setMaxSimDistance(maxViewDistance);
            save();
        }
        return maxSimDistance;
    }

    public void setMaxSimDistance(int maxSimDistance) {
        this.maxSimDistance = maxSimDistance;
    }

    public int getMinSimDistance() {
        if(minSimDistance == 0) {
            setMinSimDistance(minViewDistance);
            save();
        }
        return minSimDistance;
    }

    public void setMinSimDistance(int minSimDistance) {
        this.minSimDistance = minSimDistance;
    }

    public int getMaxChunkTickingDistance() {
        if(maxChunkTickingDistance <= 0) {
            setMaxChunkTickingDistance(8);
            save();
        }
        return maxChunkTickingDistance;
    }

    public void setMaxChunkTickingDistance(int maxChunkTickingDistance) {
        this.maxChunkTickingDistance = maxChunkTickingDistance;
    }

    public int getMinChunkTickingDistance() {
        if(minChunkTickingDistance <= 0) {
            setMinChunkTickingDistance(8);
            save();
        }
        return minChunkTickingDistance;
    }

    public void setMinChunkTickingDistance(int minChunkTickingDistance) {
        this.minChunkTickingDistance = minChunkTickingDistance;
    }

    public boolean isAllowOnClient() {
        return allowOnClient;
    }

    public void setAllowOnClient(boolean allowOnClient) {
        this.allowOnClient = allowOnClient;
    }

    public BroadcastLevel getBroadcastChangesDefault() {
        return broadcastChangesDefault;
    }

    public void setBroadcastChangesDefault(BroadcastLevel broadcastChangesDefault) {
        this.broadcastChangesDefault = broadcastChangesDefault;
    }

    public ArrayList<String> getBroadcastChanges() {
        return broadcastChanges;
    }

    public void setBroadcastChanges(ArrayList<String> broadcastChanges) {
        this.broadcastChanges = broadcastChanges;
    }

    public BroadcastLevel getBroadcastLockDefault() {
        return broadcastLockDefault;
    }

    public void setBroadcastLockDefault(BroadcastLevel broadcastLockDefault) {
        this.broadcastLockDefault = broadcastLockDefault;
    }

    public ArrayList<String> getBroadcastLock() {
        return broadcastLock;
    }

    public void setBroadcastLock(ArrayList<String> broadcastLock) {
        this.broadcastLock = broadcastLock;
    }

    public ArrayList<Rule> getRules() {
        return rules;
    }

    public void setRules(ArrayList<Rule> rules) {
        this.rules = rules;
    }

    public boolean isViewLocked() {
        return viewLocked;
    }

    public void setViewLocked(boolean viewLocked) {
        this.viewLocked = viewLocked;
    }

    public boolean isSimLocked() {
        return simLocked;
    }

    public void setSimLocked(boolean simLocked) {
        this.simLocked = simLocked;
    }

    public boolean isChunkTickingLocked() {
        return chunkTickingLocked;
    }

    public void setChunkTickingLocked(boolean chunkTickingLocked) {
        this.chunkTickingLocked = chunkTickingLocked;
    }

    @Deprecated
    public boolean deprecatedIsBroadcastToOps() {
        return deprecatedBroadcastToOps;
    }
}
